ESP32 related snippets

marekmatej opened the pull request:
Introduce snippets under `snippets/espressif` to handle configuration of
flash and PSRAM variants. This enables flexible selection of memory
variants for supported boards without proliferating board-specific
variants.

Using snippets is the right mechanism for this for several reasons:

- *Reusable across all boards:* Unlike board variants, snippets are
  global and composable. They avoid duplication by not requiring each
  board to define new variant files.
- *Supports multiple configurations for a single board:* Boards like
  `esp32s3_devkitm` can be assembled with different flash/PSRAM module
  combinations. Snippets allow these to be expressed modularly.
- *Scalable and clean:* Maintains a clean and minimal board support
  directory structure without an explosion of board variant directories.

**Why not Kconfig?**

Kconfig alone is insufficient here. While DTS symbols can be referenced
from Kconfig, conditional DTS inclusion *based on Kconfig options* is
not supported. We need the ability to conditionally add devicetree
overlays and configuration fragments, which snippets provide cleanly.

This approach aligns with the modular and composable philosophy of the
Zephyr build system.

The following configurations have been added:

- *Flash size* with appropriate partition selection
- *PSRAM size* and modes configuration 
- *PSRAM reloc* allows to relocate the flash code and rodata into a PSRAM
- *PSRAM wifi* related configuration cases

The goal is to provide a unified and simple way to maintain memory configuration that is frequently needed in development.
`west build -b <esp32s3_devkitm/esp32s3/procpu samples/net/wifi/shell -p -S "flash-32M psram-8M psram-reloc"`

All Espressif boards documentation has been updated to cover the board variant options using snippets.

marekmatej committed:
snippets: add Espressif related snippets

Introduces the flash and PSRAM variants suitable for every ESP32 related
board.

Signed-off-by: Marek Matej <marek.matej@espressif.com>

tejlmand commented on snippets/espressif-mcuboot/esp-mcuboot.conf:
```diff
@@ -0,0 +1 @@
+CONFIG_BOOTLOADER_MCUBOOT=y
```
Why is this a snippet ?

Please explain why `-DSNIPPET=esp-mcuboot` is cleaner, better, easier than `-DCONFIG_BOOTLOADER_MCUBOOT=y`

marekmatej replied:
Yeah, the idea here is to use them in combination. I'm using `-DSNIPPET="mcuboot;spiram"`, which is much shorter compared to adding `-DCONFIG_BOOTLOADER_MCUBOOT=y`

marekmatej replied:
> Why is this a snippet ?
> 
> Please explain why `-DSNIPPET=esp-mcuboot` is cleaner, better, easier than `-DCONFIG_BOOTLOADER_MCUBOOT=y`

We often need to switch between no-bootloader (simpleboot) to MCUBoot. And this - yes very primitive - mcuboot snippet allows us to use snippets in a linear fashion:
`west build -p -b <board> <sample> -DSNIPPER="mcuboot;spiram;..."`

tejlmand commented on snippets/espressif-mcuboot/snippet.yml:
If we want a snippet for enabling MCUBoot, then why is this an espressif specific snippet, and not just a MCUboot snippets which can do board/SoC specific stuff if that is needed ?

marekmatej replied:
We are currently in the process of integrating the MCUboot Espressif Port into Zephyr, and it seems there will be additional configs involved

tejlmand replied:
then I prefer we wait with a MCUboot specific snippet until we have a better view of what such snippet will actually do, and whether such snippet should have esp in it's name, or be a generic mcuboot snippet which can have special handling for esp SoCs.

marekmatej replied:
I agree with you, and have renamed the `mcuboot` snippet files and identifiers.

tejlmand commented:
> I like the `mcuboot` snippet because it's shorter and allows sharing the image features as a single string.

but following such approach to its extreme would mean that any single Kconfig could end-up having a corresponding snippet.
What about the next snippet which simply enables just a single Kconfig, should that also be accepted ?

marekmatej commented:
> > I like the `mcuboot` snippet because it's shorter and allows sharing the image features as a single string.
> 
> but following such approach to its extreme would mean that any single Kconfig could end-up having a corresponding snippet. What about the next snippet which simply enables just a single Kconfig, should that also be accepted ?

Hypothetically, but we are talking about enabling bootloader support. It is rather exception than a rule. Frankly, creating 3 new files just to enable single config is not a victory. But it makes perfect sense in this case.

marekmatej commented on snippets/espressif-spiram-8M/esp-spiram-8M.conf:
```diff
@@ -0,0 +1,6 @@
+# SPDX-License-Identifier: Apache-2.0
+
+CONFIG_ESP_SPIRAM=y
+CONFIG_ESP_SPIRAM_SIZE=8388608
```
> Kconfig is for software configuration. RAM size, which is a non-configurable hardware characteristic, should come from devicetree.

Nope, the PSRAM size is the same problem as flash size. We have 4ex `esp32s3_devkitc`, which is often assembled with various combinations of PSRAM size and flash sizes.

JarmouniA reviewed (commented):
As a minimum, the documentation of the relevant ESP32 boards should be updated to indicate the availability of these snippets and how they can be used to switch between the different board variants.

marekmatej commented:
> From the PR description
> 
> > Using snippets is the right mechanism for this for several reasons:
> > 
> > * Supports multiple configurations for a single board:
> 
> Those are fixed hardware configurations, and external Flash/RAM sizes are part of a board's unchangeable characteristics. When those sizes change, it's no longer the SAME board, it's a variant.
> 
> > Boards like esp32s3_devkitm can be assembled with different flash/PSRAM module combinations.
> 
> They are assembled in the factory, not by the user, the external PSRAM/Flash are not added using a shield for ex.

Yet, every one of us has multiple `esp32s3_devkitc`, each assembled with a different module with various flash and PSRAM (+ PSRAM mode). They are all made by Espressif.
As much as I agree that is a *variant*, you need to alter every ESP32 board files to implement the variants. Even if you do so, you still have not addressed the PSRAM mode.

marekmatej commented:
> As a minimum, the documentation of the relevant ESP32 boards should be updated to indicate the availability of these snippets and how they can be used to switch between the different board variants.

Ok, that could be added to the PR.

marekmatej committed:
boards: espressif: Update boards documentation

Added a new 'Board variants' section to the board documentation
to describe supported flash and PSRAM configurations for Espressif
boards using snippet-based build-time selection.

Signed-off-by: Marek Matej <marek.matej@espressif.com>

marekmatej commented on snippets/espressif/psram-reloc/README.rst:
```diff
@@ -0,0 +1,9 @@
+Overview
+********
+
+This snippet allows users to enable the flash section to be relocated to PSRAM.
```
There is no "internal flash" in the ESP32, only SPI flash. I'll leave this as is.

marekmatej replied:
> But ESP32S3-WROOM for ex. has 384 KB ROM.

Every ESP32 has ROM with loader code, etc., but it is not subject to mapping or relocation. You can call ROM functions, and avoid interfering with ROM-code data. That's all :)

marekmatej commented on boards/espressif/esp32_devkitc/doc/index.rst:
```diff
@@ -190,6 +190,44 @@ message in the monitor:
+==============
+
+ESP32 boards can be assembled with different modules using multiple combinations of flash sizes, PSRAM sizes and PSRAM modes.
+Snippets provide a modular way to apply these variations at build time without duplicating board definitions.
```
Thanks, this formulation seems much better to me.

kartben reviewed (dismissed):
I think it's high time you start using Sphinx `.. include::` statements in the doc pages, it just doesn't scale to copy-paste over and over :) As an alternative, you may want to host some generic "tips and tricks" underneath boards/espressif and link relevant sections from the boards' pages as needed.

marekmatej commented:
> I think it's high time you start using Sphinx `.. include::` statements in the doc pages, it just doesn't scale to copy-paste over and over :) As an alternative, you may want to host some generic "tips and tricks" underneath boards/espressif and link relevant sections from the boards' pages as needed.

You're right, I'll move the chapter into a `boards/espressif/board-variants.rst`

marekmatej commented on snippets/espressif/psram-2M/esp-psram-2M.conf:
```diff
@@ -0,0 +1,6 @@
+# SPDX-License-Identifier: Apache-2.0
+
+CONFIG_ESP_SPIRAM=y
+CONFIG_ESP_SPIRAM_SIZE=2097152
```
I think we might be spinning in circles. The change you are suggesting boils down to the same `CONFIG_ESP_SPIRAM_SIZE=2097152`. Could you elaborate a little bit more on the topic of possible hidden bugs?

However, I took your recommendation into account and re-worked the PR accordingly. It is kind of better as it uses the device tree instead of hard config. PTAL at changes and approve.
